### ISSUE TITLE
[BUG] job information was replaced by the output of one of its states

### ISSUE THREAD

**Issue by erwindon** [saltstack/salt#58968]
**Description**
Running a `state.highstate` where one of the tasks performs a `salt.function` with `cmd.run` with `ls -l`.
The whole output of this job is replaced by data from the `ls -l` (actually only the first column from it).
Thereby completely destroying the regular output that was (likely) also returned.

This is not easily repeatable, so I suspect a race condition of some kind.

**Setup**
configure a state `salt.function` with arguments `cmd.run` and `ls -l`.
configure some other states as well.

**Steps to Reproduce the behaviour**
run `state.apply`.
The expected return value has a list of minions, each with a list of tasks.
In this case, the output from `jobs.list_job` did not contain a list of minions and their tasks as expected.
Instead, it was filled with output from `ls`:
```json
{
    "return": [
        {
            "jid": "20201117213929593188",
            "Function": "bbb.txt",
            "Arguments": [
                "ccc.txt",
                "ddd.txt",
                "eee.txt",
                "state.apply"
            ],
            "Target": "aaa.txt",
            "Target-type": "glob",
            "User": "sudo_erwindon",
            "StartTime": "2020, Nov 17 21:39:29.593188",
            "Result": {}
        }
    ]
}
```
Note that the first filename is placed in `Target`, the second one in `Function` and the remaining ones and an extra `state.apply` in `Arguments`.
(the actual filenames that were listed are redacted).

**Expected behaviour**
A regular state output of `state.apply` with the list of minions and their states.

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
Salt Version:
           Salt: 3002.1
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: unknown
       dateutil: 2.7.3
      docker-py: 3.4.1
          gitdb: 2.0.5
      gitpython: 2.1.11
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.6.1
         pygit2: Not Installed
         Python: 3.7.3 (default, Jul 25 2020, 13:03:44)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.1.2
          smmap: 2.0.5
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.1
 
System Versions:
           dist: debian 10 buster
         locale: utf-8
        machine: x86_64
        release: 4.19.0-12-amd64
         system: Linux
        version: Debian GNU/Linux 10 buster
```
</details>

**Comment by frogunder** [comment]
I have not been able to replicate this. 
@saltstack/team-core can either of you take a look at this when you have time? Thanks.

**Comment by waynew** [comment]
I tried this a few times to no avail (in fact, with the provided information I can't even get salt.function to run the way it's described, I only get

```
Salt request timed out. The master is not responding. You may need to run your command with `--async` in order to bypass the congested event bus. With `--async`, the CLI tool will print the job id (jid) and exit immediately without listening for responses. You can then use `salt-run jobs.lookup_jid` to look up the results of the job in the job cache later.
```
)

I'm going to go ahead and close for now, since it doesn't appear that anyone can reproduce it. It's definitely a weird bug, and TBH I'm not entirely sure how it would even occur -- I mean I have some theories, but they're pretty weak theories. I feel like the most likely scenario is actually just one involving caching... though it's *possible* that when building up some return values a dict got merged incorrectly.

The thing that seems weirdest to me is the whole salt.function/cmd.run part. But again, since we can't appear to reproduce this, we'll go ahead and close until it either surfaces again or we can actually reproduce.
